File: searchnsort/ltcd_2411.py
'''
LeetCode - problem 2441
'''

    def smallestSubarrays(self, nums: List[int]) -> List[int]:
        maximum_bits = int(log(10**9, 2)) + 1
        index_for_each_bit = [[] for _ in range(maximum_bits)]
        for i in range(len(nums)):
            n = nums[i]
            for bit in range(maximum_bits):
                if n % 2 == 1:
                    index_for_each_bit[bit].append(i)
                n >>= 1

        res = []
        for i in range(len(nums)):
            cur_idx = [i]
            n = nums[i]
            for bit in range(maximum_bits):
                bit_index = index_for_each_bit[bit]
                if n % 2 == 0:
                    k = bisect_left(bit_index, i)
                    if k < len(bit_index):
                       cur_idx.append(bit_index[k])
                n >>= 1
            res.append(max(cur_idx) - i + 1)
        return res

File: searchnsort/ltcd_2411.cpp
//LeetCode - problem 2411

    vector<int> smallestSubarrays(vector<int>& nums) {
        int sz=nums.size();
        vector<vector<int>>arr(31);
        int j=0;
        for(auto & ele:nums){
            for(int i=0;i<31;i++){
                if((ele>>i)&1)
                    arr[i].push_back(j);
            }
            j++;
        }
        vector<int>suf=nums;
        for(int i=sz-2;i>=0;i--)
            suf[i]=nums[i]|suf[i+1];
        vector<int>ans(sz,0);
        for(int i=0;i<sz;i++){
	    int res=1;
            for(int k=0;k<31;k++){
                if((suf[i]>>k)&1){
                    auto it=lower_bound(begin(arr[k]),end(arr[k]),i);
                    int len=(*it)-i+1;
                    res=max(res,len);
                }
            }
            ans[i]=res;
        }
        return ans;
    }
